driver/src/main/com/mongodb/reactivestreams/client/internal/GridFSDownloadPublisherImpl.java

```diff
@@ -0,0 +1,214 @@
+public class GridFSDownloadPublisherImpl implements GridFSDownloadPublisher {
+    private final com.mongodb.reactivestreams.client.gridfs.GridFSDownloadStream gridFSDownloadStream;
+    private int bufferSizeBytes;
+
+    GridFSDownloadPublisherImpl(final com.mongodb.reactivestreams.client.gridfs.GridFSDownloadStream gridFSDownloadStream) {
+        this.gridFSDownloadStream = gridFSDownloadStream;
+    }
+
+    @Override
+    public Publisher<GridFSFile> getGridFSFile() {
+        return gridFSDownloadStream.getGridFSFile();
+    }
+
+    @Override
+    public GridFSDownloadPublisher bufferSizeBytes(final int bufferSizeBytes) {
+        this.bufferSizeBytes = bufferSizeBytes;
+        return this;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super ByteBuffer> s) {
+        s.onSubscribe(new GridFSDownloadSubscription(s));
+    }
+
+    class GridFSDownloadSubscription implements Subscription {
+        private final Subscriber<? super ByteBuffer> outerSubscriber;
+        private final Object lock = new Object();
+
+        /* protected by `lock` */
+        private GridFSFile gridFSFile;
+        private boolean requestedData;
+        private long sizeRead = 0;
+        private long requested = 0;
+        private int currentBatchSize = 0;
+        private boolean isCompleted = false;
+        private boolean isTerminated = false;
+        /* protected by `lock` */
+
+        public GridFSDownloadSubscription(final Subscriber<? super ByteBuffer> outerSubscriber) {
+            this.outerSubscriber = outerSubscriber;
+        }
+
+        private final Subscriber<GridFSFile> gridFSFileSubscriber = new Subscriber<GridFSFile>() {
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final GridFSFile result) {
+                synchronized (lock) {
+                    gridFSFile = result;
+                }
+            }
+
+            @Override
+            public void onError(final Throwable t) {
+                outerSubscriber.onError(t);
+                terminate();
+            }
+
+            @Override
+            public void onComplete() {
+                synchronized (lock) {
+                    requestedData = false;
+                }
+                requestMoreOrComplete();
+            }
+        };
+
+        class GridFSDownloadStreamSubscriber implements Subscriber<Integer> {
+            private final ByteBuffer byteBuffer;
+
+            GridFSDownloadStreamSubscriber(final ByteBuffer byteBuffer) {
+                this.byteBuffer = byteBuffer;
+            }
+
+            @Override
+            public void onSubscribe(final Subscription s) {
+                s.request(1);
+            }
+
+            @Override
+            public void onNext(final Integer integer) {
+                synchronized (lock) {
+                    sizeRead += integer;
```

Comment:
Tracks the size of the data read, so we know when to complete the download publisher.